kernel/src/table_features/variant_type.rs

```diff
@@ -0,0 +1,158 @@
+//! Validation for variantType and variantType-preview feature support
+
+use super::{ReaderFeature, WriterFeature};
+use crate::actions::Protocol;
+use crate::schema::{PrimitiveType, Schema, SchemaTransform};
+use crate::utils::require;
+use crate::{DeltaResult, Error};
+
+use std::borrow::Cow;
+
+pub(crate) fn validate_variant_type_feature_support(
+    schema: &Schema,
+    protocol: &Protocol,
+) -> DeltaResult<()> {
+    // Both the reader and writer need to have either the VariantType or the VariantTypePreview
+    // features.
+    if (!protocol.has_reader_feature(&ReaderFeature::VariantType)
+        && !protocol.has_reader_feature(&ReaderFeature::VariantTypePreview)) ||
+        (!protocol.has_writer_feature(&WriterFeature::VariantType)
+        && !protocol.has_writer_feature(&WriterFeature::VariantTypePreview))
+    {
+        let mut uses_variant = UsesVariant(false);
+        let _ = uses_variant.transform_struct(schema);
+        require!(
+            !uses_variant.0,
+            Error::unsupported(
+                "Table contains VARIANT columns but does not have the required 'variantType' feature in reader and writer features"
+            )
+        );
+    }
+    Ok(())
+}
+
+/// Schema visitor that checks if any column in the schema uses VARIANT type
+pub struct UsesVariant(pub bool);
```

Comment:
(and if you derive `Default` you can just do `UsesVariant::default()` at callsites
```suggestion
#[derive(Debug, Default)]
pub(crate) struct UsesVariant(pub(crate) bool);
```